istioctl/cmd/workload.go

```diff
@@ -0,0 +1,402 @@
+	}
+	if err := ioutil.WriteFile(filepath.Join(dir, "istio-token"), []byte(tokenReq.Status.Token), filePerms); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO: Support the proxy.istio.io/config annotation
+func createMeshConfig(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, clusterID, revision, dir string) (*meshconfig.MeshConfig, error) {
+	istioCM := "istio"
+	// Case with multiple control planes
+	if revision != "" {
+		istioCM = fmt.Sprintf("%s-%s", istioCM, revision)
+	}
+	istio, err := kubeClient.CoreV1().ConfigMaps(istioNamespace).Get(context.Background(), istioCM, metav1.GetOptions{})
+	// errors if the requested configmap does not exist in the given namespace
+	if err != nil {
+		return nil, fmt.Errorf("configmap %s was not found in namespace %s: %v", istioCM, istioNamespace, err)
+	}
+	meshConfig, err := mesh.ApplyMeshConfigDefaults(istio.Data[configMapKey])
+	if err != nil {
+		return nil, err
+	}
+
+	labels := map[string]string{}
+	for k, v := range wg.Spec.Metadata.Labels {
+		labels[k] = v
+	}
+	// case where a user provided custom workload group has labels in the workload entry template field
+	for k, v := range wg.Spec.Template.Labels {
+		labels[k] = v
+	}
+	we := wg.Spec.Template
+	md := meshConfig.DefaultConfig.ProxyMetadata
+	md["CANONICAL_SERVICE"], md["CANONICAL_REVISION"] = inject.ExtractCanonicalServiceLabels(labels, wg.Name)
+	md["DNS_AGENT"] = ""
+	md["POD_NAMESPACE"] = wg.Namespace
+	md["SERVICE_ACCOUNT"] = we.ServiceAccount
+	md["TRUST_DOMAIN"] = meshConfig.TrustDomain
+
+	md["ISTIO_META_CLUSTER_ID"] = clusterID
+	md["ISTIO_META_MESH_ID"] = meshConfig.DefaultConfig.MeshId
+	md["ISTIO_META_NETWORK"] = we.Network
+	if portsJSON, err := json.Marshal(we.Ports); err == nil {
+		md["ISTIO_META_POD_PORTS"] = string(portsJSON)
+	}
+	md["ISTIO_META_WORKLOAD_NAME"] = wg.Name
+	labels["service.istio.io/canonical-name"] = md["CANONICAL_SERVICE"]
+	labels["service.istio.io/canonical-version"] = md["CANONICAL_REVISION"]
+	if labelsJSON, err := json.Marshal(labels); err == nil {
+		md["ISTIO_METAJSON_LABELS"] = string(labelsJSON)
+	}
+
+	meshYAML, err := gogoprotomarshal.ToYAML(meshConfig)
+	if err != nil {
+		return nil, err
+	}
+	return meshConfig, ioutil.WriteFile(filepath.Join(dir, "mesh"), []byte(meshYAML), filePerms)
+}
+
+// Retrieves the external IP of the ingress-gateway for the hosts file additions
+func createHosts(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, meshConfig *meshconfig.MeshConfig, dir string) error {
+	ingress, err := kubeClient.CoreV1().Services(istioNamespace).Get(context.Background(), multicluster.IstioIngressGatewayServiceName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("service %s was not found in namespace %s: %v", multicluster.IstioIngressGatewayServiceName, istioNamespace, err)
+	}
+	serviceIP := ingress.Status.LoadBalancer.Ingress[0].IP
```

Comment:
On a cluster without load balancer:

```
$ ./ik x workload entry configure --name foo --namespace bar -o /tmp/foo --clusterID=Kubernetes
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
istio.io/istio/istioctl/cmd.createHosts(0x3315ee0, 0xc000e8db00, 0xc000f6eb60, 0xc00000c5a0, 0x7ffcfc669971, 0x8, 0x0, 0x7ffcfc669971)
        /home/howardjohn/go/src/istio.io/istio/istioctl/cmd/workload.go:386 +0x731
istio.io/istio/istioctl/cmd.createConfig(0x3315ee0, 0xc000e8db00, 0xc000f6eb60, 0x7ffcfc669986, 0xa, 0x0, 0x0, 0x7ffcfc669971, 0x8, 0x0, ...)
        /home/howardjohn/go/src/istio.io/istio/istioctl/cmd/workload.go:240 +0x1dd
istio.io/istio/istioctl/cmd.configureCommand.func2(0xc000e3c2c0, 0xc000daf570, 0x0, 0x7, 0x0, 0x0)
        /home/howardjohn/go/src/istio.io/istio/istioctl/cmd/workload.go:195 +0x17f
github.com/spf13/cobra.(*Command).execute(0xc000e3c2c0, 0xc000daf500, 0x7, 0x7, 0xc000e3c2c0, 0xc000daf500)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0xc000e72840, 0xb, 0xb, 0xc000e72840)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
        /home/howardjohn/go/src/istio.io/istio/istioctl/cmd/istioctl/main.go:36 +0x126
```

Reply:
I thought that every cluster would have a load balancer, not sure where else the ingress IP would be exposed in the case it doesn't

Reply:
It probably does, but we should gracefully handle it not panic